File: src/client/utils/images.cpp
/*
   Drawpile - a collaborative drawing program.

   Copyright (C) 2014 Calle Laakkonen

   Drawpile is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   Drawpile is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with Drawpile.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "images.h"

#include <QSize>
#include <QImageWriter>
#include <QImage>
#include <QColor>

namespace utils {

//! Check if image dimensions are not too big. Returns true if size is OK
bool checkImageSize(const QSize &size)
{
	// The protocol limits width and height to 2^29 (PenMove, 2^32 for others)
	// However, QPixmap can be at most 32767 pixels wide/tall
	static const int MAX_SIZE = 32767;

	// Rather than limiting the image dimensions to a safe square, allow large rectangular images
	// by limiting size by required memory
	static const quint64 MAX_MEM = 1024 * 1024 * 1024;
	static const quint64 MAX_PIXELS = MAX_MEM / 4;

	return
		size.width() <= MAX_SIZE &&
		size.height() <= MAX_SIZE &&
		quint64(size.width())*quint64(size.height()) <= MAX_PIXELS;
}

bool isWritableFormat(const QString &filename)
{
	const int dot = filename.lastIndexOf('.');
	if(dot<0)
		return false;
	const QByteArray suffix = filename.mid(dot+1).toUtf8();

	for(const QPair<QString,QByteArray> &fmt : writableImageFormats()) {
		if(fmt.second == suffix)
			return true;
	}

	return false;
}

QList<QPair<QString,QByteArray>> writableImageFormats()
{
	QList<QPair<QString,QByteArray>> formats;

	// We support ORA ourselves
	formats.append(QPair<QString,QByteArray>("OpenRaster", "ora"));

	// Get list of available formats
	for(const QByteArray &fmt : QImageWriter::supportedImageFormats())
	{
		// only offer a reasonable subset
		if(fmt == "png" || fmt=="jpeg" || fmt=="bmp" || fmt=="gif" || fmt=="tiff")
			formats.append(QPair<QString,QByteArray>(QString(fmt).toUpper(), fmt));
		else if(fmt=="jp2")
			formats.append(QPair<QString,QByteArray>("JPEG2000", fmt));
	}
	return formats;
}

QColor isSolidColorImage(const QImage &image)
{
	Q_ASSERT(image.format() == QImage::Format_ARGB32);
	if(image.format() != QImage::Format_ARGB32) {
		qWarning("isSolidColorImage: not an ARGB32 image!");
		return QColor();
	}

	const quint32 c = *reinterpret_cast<const quint32*>(image.bits());
	const quint32 *p = reinterpret_cast<const quint32*>(image.bits());

	int len = image.width() * image.height();
	while(--len) {
		if(*(++p) != c)
			return QColor();
	}
	return QColor::fromRgba(c);
}

}

File: src/dpcore/src/canvas/images.rs
// This file is part of Drawpile.
// Copyright (C) 2021 Calle Laakkonen
//
// Drawpile is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// As additional permission under section 7, you are allowed to distribute
// the software through an app store, even if that store has restrictive
// terms and conditions that are incompatible with the GPL, provided that
// the source is also available under the GPL with or without this permission
// through a channel without those restrictive terms and conditions.
//
// Drawpile is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Drawpile.  If not, see <https://www.gnu.org/licenses/>.

use super::compression::ImageCompressor;
use crate::paint::{Blendmode, Image, LayerID, Rectangle, UserID};
use crate::protocol::message::{CommandMessage, PutImageMessage};

const MAX_IMAGE_LEN: usize = 0xffff - 19;


/// Make a PutImage command.
///
/// If the image is too big to fit into a single command,
/// multiple PutImages will be written.
pub fn make_putimage(
    user: UserID,
    layer: LayerID,
    x: u32,
    y: u32,
    image: &Image,
    mode: Blendmode,
) -> Vec<CommandMessage> {
    let mut messages = Vec::new();
    make_putimage_crop(
        &mut messages,
        user,
        layer,
        x,
        y,
        image,
        mode,
        &Rectangle::new(0, 0, image.width as i32, image.height as i32)
    );
    messages
}

fn make_putimage_crop(
    messages: &mut Vec<CommandMessage>,
    user: UserID,
    layer: LayerID,
    x: u32,
    y: u32,
    image: &Image,
    mode: Blendmode,
    rect: &Rectangle,
) {
    // Try compressing this subregion of the image
    {
        let mut compressor = ImageCompressor::new();
        image.rect_iter(rect).for_each(|row| compressor.add(row));
        let compressed = compressor.finish();

        if compressed.len() <= MAX_IMAGE_LEN {
            messages.push(CommandMessage::PutImage(
                user,
                PutImageMessage {
                    layer,
                    mode: mode.into(),
                    x: x + rect.x as u32,
                    y: y + rect.y as u32,
                    w: rect.w as u32,
                    h: rect.h as u32,
                    image: compressed,
                },
            ));
            return;
        }
    }

    // Image too big to fit into a message. Split into four pieces and
    // try again
    // TODO: a smarter approach would be to split at tile boundaries to
    // make the PutImage operations more efficient
    let splitx = rect.w / 2;
    let splity = rect.h / 2;

    make_putimage_crop(
        messages,
        user,
        layer,
        x,
        y,
        image,
        mode,
        &Rectangle::new(rect.x, rect.y, splitx, splity),
    );
    make_putimage_crop(
        messages,
        user,
        layer,
        x,
        y,
        image,
        mode,
        &Rectangle::new(
            rect.x + splitx,
            rect.y,
            rect.w - splitx,
            splity,
        ),
    );
    make_putimage_crop(
        messages,
        user,
        layer,
        x,
        y,
        image,
        mode,
        &Rectangle::new(
            rect.x,
            rect.y + splity,
            splitx,
            rect.h - splity,
        ),
    );
    make_putimage_crop(
        messages,
        user,
        layer,
        x,
        y,
        image,
        mode,
        &Rectangle::new(
            rect.x + splitx,
            rect.y + splity,
            rect.w - splitx,
            rect.h - splity,
        ),
    );
}
